### PR TITLE
Pi-hole Core v5.1

### LINKED ISSUE #2888: Installer exits with php5-* packages chosen on Buster
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`9`

---
**Expected behaviour:**

`On Debian/Ubuntu, before checking for any packages on repos (php vs php5, sqlite vs sqlite3 etc), an APT update should be done to assure that lists are available and up-to-date.`

**Actual behaviour:**

`Package availability is done first, so choices can be done on wrong assumptions, especially since no availability check is done on the fallback packages anymore.`

**Steps to reproduce:**

1. Clear the APT cache + package lists
2. Run Pi-hole installer
3. Watch it choose php5-* packages, because dry-run install of `php` fails. But php5-* packages are not available as well, thus installer fails.

**Solution:**

- Run APT update step before checking the repo for available packages.
- Check fallback package availability as well, if those fail as well, exit installer with meaningful error message.
- Optional: Use `apt-cache show <pkg>` to check for package availability, which is much faster and cleaner then a dry-run install.

### LINKED ISSUE #3217: pihole installation fails if group pihole already exists
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`{10}`

The issue is in `create_pihole_user()` in `basic-install.sh` not checking first if the group `pihole` exists. An additional if statement should be added to utilise an alternate `useradd` command such as `useradd -r -s /usr/sbin/nologin -g pihole pihole`

---
**Expected behaviour:**

Pihole successfully installing

**Actual behaviour:**

Pihole fails to create user pihole and then installation fails when attempting to access non-existent gravity.sh

```
zander@zpi:~/Pi-hole/automated install$ sudo bash basic-install.sh

  [✓] Root user check

        .;;,.
        .ccccc:,.
         :cccclll:.      ..,,
          :ccccclll.   ;ooodc
           'ccll:;ll .oooodc
             .;cll.;;looo:.
                 .. ','.
                .',,,,,,'.
              .',,,,,,,,,,.
            .',,,,,,,,,,,,....
          ....''',,,,,,,'.......
        .........  ....  .........
        ..........      ..........
        ..........      ..........
        .........  ....  .........
          ........,,,,,,,'......
            ....',,,,,,,,,,,,.
               .',,,,,,,,,'.
                .',,,,,,'.
                  ..'''.

  [i] Existing PHP installation detected : PHP version 7.2.24-0ubuntu0.18.04.3
  [✓] Disk space check
  [✓] Update local cache of available packages

  [✓] Checking apt-get for upgraded packages... up to date!

  [i] Installer Dependency checks...
  [✓] Checking for apt-utils
  [✓] Checking for dialog
  [✓] Checking for debconf
  [✓] Checking for dhcpcd5
  [✓] Checking for git
  [✓] Checking for iproute2
  [✓] Checking for whiptail

  [i] Using interface: eth0
  [i] Using   [i] Static IP already configured
  [i] Found IPv6 GUA address, using it for blocking IPv6 ads
  [i] IPv4 address: 192.168.124.7/24
  [i] IPv6 address: 2c0f:ef18:2410:0:d104:44df:86f5:5747
  [i] Web Interface On
  [i] Web Server On
  [i] Logging On.
  [✓] Check for existing repository in /etc/.pihole
  [✓] Update repo in /etc/.pihole

  [✓] Check for existing repository in /var/www/html/admin
  [✓] Update repo in /var/www/html/admin

  [i] Main Dependency checks...
  [✓] Checking for cron
  [✓] Checking for curl
  [✓] Checking for dnsutils
  [✓] Checking for iputils-ping
  [✓] Checking for lsof
  [✓] Checking for netcat
  [✓] Checking for psmisc
  [✓] Checking for sudo
  [✓] Checking for unzip
  [✓] Checking for wget
  [✓] Checking for idn2
  [✓] Checking for sqlite3
  [✓] Checking for libcap2-bin
  [✓] Checking for dns-root-data
  [✓] Checking for resolvconf
  [✓] Checking for libcap2
  [✓] Checking for lighttpd
  [✓] Checking for php7.2-common
  [✓] Checking for php7.2-cgi
  [✓] Checking for php7.2-sqlite3

  [✓] Enabling lighttpd service to start on reboot...
  [i] Creating user 'pihole'...useradd: group pihole exists - if you want to add this user to that group, use -g.
  [✗] Creating user 'pihole'

  [i] FTL Checks...

  [✓] Detected ARM-aarch64 architecture
  [i] Checking for existing FTL binary...
  [i] Latest FTL Binary already installed (v4.3.1). Confirming Checksum...
  [i] Checksum correct. No need to download!
  [i] Creating user 'pihole'...useradd: group pihole exists - if you want to add this user to that group, use -g.
  [✗] Creating user 'pihole'
usermod: user 'pihole' does not exist
grep: /etc/pihole/setupVars.conf: No such file or directory
  [i] Testing if systemd-resolved is enabled
  [i] Systemd-resolved does not need to be restarted
  [✓] Restarting lighttpd service...
  [✓] Enabling lighttpd service to start on reboot...
  [i] Restarting services...
  [✓] Enabling pihole-FTL service to start on reboot...
  [✓] Restarting pihole-FTL service...
basic-install.sh: line 1724: /opt/pihole/gravity.sh: No such file or directory
```

**Steps to reproduce:**

Try to install pihole while the group `pihole` already exists

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

N/A

**Troubleshooting undertaken, and/or other relevant information:**

`groupdel pihole`

The group `pihole` existed from a previous pihole installation I removed. Not sure if the uninstaller failed to remove the group as well as the user `pihole` or it may be an oversight in the uninstaller

### LINKED ISSUE #3336: Installation stops if no blocklist is selected. v5.0
**In raising this issue, I confirm the following:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`1`

---
**Expected behaviour:**
Doing a fresh install of PiHole 5.0, if we do not select any list in the initial setup, installation should continue, because we will add our custom list later.

**Actual behaviour:**

If we do not select any blocklist in the initial setup step, installation fails because `/etc/pihole/adlists.list` do not exist

Captura

Install log:
```
user-01@srv-01:~$ curl -sSL https://install.pi-hole.net | sudo bash
[sudo] password for user-01: 

  [✓] Root user check

        .;;,.
        .ccccc:,.
         :cccclll:.      ..,,
          :ccccclll.   ;ooodc
           'ccll:;ll .oooodc
             .;cll.;;looo:.
                 .. ','.
                .',,,,,,'.
              .',,,,,,,,,,.
            .',,,,,,,,,,,,....
          ....''',,,,,,,'.......
        .........  ....  .........
        ..........      ..........
        ..........      ..........
        .........  ....  .........
          ........,,,,,,,'......
            ....',,,,,,,,,,,,.
               .',,,,,,,,,'.
                .',,,,,,'.
                  ..'''.

  [i] Existing PHP installation detected : PHP version 7.4.5
  [✓] Disk space check
  [✓] Update local cache of available packages

  [✓] Checking apt-get for upgraded packages... 69 updates available
  [i] It is recommended to update your OS after installing the Pi-hole!

  [i] Installer Dependency checks...
  [✓] Checking for dhcpcd5
  [✓] Checking for git
  [✓] Checking for iproute2
  [✓] Checking for whiptail

  [i] SELinux not detected
  [i] Using upstream DNS: Cloudflare (1.1.1.1, 1.0.0.1)
chmod: cannot access '/etc/pihole/adlists.list': No such file or directory error when the file exists
```

**Steps to reproduce:**
Simply make a fresh install and leave initial blocklists unselected.

**Troubleshooting undertaken, and/or other relevant information:**
By selecting at least one list, the installation continues normally.

### LINKED ISSUE #3444: gravity.sh blocklist URL regex should allow semicolons
After some discussion on Reddit:

 * https://www.reddit.com/r/pihole/comments/gqp30x/i_just_updated_to_50_and_im_trying_to_do_a_pihole/frv2yla/?context=3

It seems that the script doesn't consider a semicolon (`;`) valid in blocklist URLs. My list uses them by defaults as query string parameter delimiters, eg:

 * https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts;showintro=0

Relevant line is here:

 * https://github.com/pi-hole/pi-hole/blob/4d25f695267590b61a4061f9bb43448005d99b85/gravity.sh#L379

### PR DESCRIPTION
**Pull request by PromoFaux**
## New
- Improve conditional forwarding #3260 (@DL6ER)
- Added support for CNAME record add/remove #3351 (@marank)
- Implement backend support for theming #3327 (@DL6ER)
- Debug extras #3423 (@PromoFaux)
- Add CAP_SYS_NICE to FTL's capabilities #3455 (@DL6ER)
- Display warning if unsupported OS is detected #3441 (Subsquently fixed by #3490) (@PromoFaux)

## Tweaks
- Remove wget from PIHOLE_DEPS #3185 (@msamendinger)
- use py3 instead py2 #3153 (@jnozsc)
- Assorted typo fixes #3126 (@XhmikosR)
- Reduce apt-get install verbosity #3205 (@MichaIng)
- fix typo #3218 (@JulienOrain)
- Revert using hardcoded php-intl meta package #3204 (@MichaIng)
- Do not add pihole user to web server group #3226 (@MichaIng)
- Remove dead code #3312 (@subdavis}
- Only check SELinux states if "getenforce" command exists #3353 (@MichaIng)
- Update documentation links #3359 (@dnicolson)
- Fix GitHub capitalisation #3360 (@dnicolson)
- Remove store/shop link from README #3369 (@a-p-o)
- Start pihole-FTL after network #3248 (@MichaIng)
- Use `toggle` instead of `select` #3406 (@PromoFaux)
- Remove csslint from Strickler CI #3419 (@XhmikosR)
- Reload the lists, which will cause the cached privacy level to be updated #3415 (@fhriley)
- Remove deprecated Disconnect lists from "defaults" #3451 (@PromoFaux )
- Update links #3412 (@XhmikosR)
- Allow certain domain duplicates #3466  (@DL6ER )
- Change OS Detection in debug script #3434 (@PromoFaux)
- Improve man pihole restartdns section #3492 (@DL6ER)
- Improve IP validation function #3489 (@Promofaux)
- Markdown tweaks #3420 (@XhmikosR)
- Fix/tweak blocking page #3403 (@XhmikosR)
- Tweak lighttpd configs #3425 (@XhmikosR)

## Fixes
- Minor bugfix for Binary detection #3347 (@PromoFaux)
- fixes #3217 by checking for existing pihole group #3246 (@pvogt09)
- Failsafe checks for available APT package #3203 - Fixes #2888 (@MichaIng)
- fix #3336 by creating adlist file even if no list was selected by user #3388 (@pvogt09)
- Updated gravity.sh to allow semicolons in URLs #3446 - Fixes #3444 (@pgl)
- Create adlist file even if no options are selected on install #3463 (@jesterchen)
- Replace possible "\#" by "#" to fix AdminLTE#1427 #3470 (@DL6ER)

### CODE REVIEW

**Comment by dschaper** [comment]
We ain't getting 3 approvals kids...

**Comment by DL6ER** [comment]
So we need to charge the rules

**Comment by PromoFaux** [comment]
Dropped it down to requiring two only

File: supportedos.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Raspbian=9,10
+Ubuntu=16,18
```
Comment (MichaIng) [inline supportedos.txt:2]: You are quite strict on Ubuntu, supporting a two years old version at max? Consider that 20.04 LTS has been released (and their half-year release schedule in general), so I bet you'll face many irritated user reports with this when it leads to an installer exit now, requiring to set an environment variable to pass.

However, matches: https://docs.pi-hole.net/main/prerequisites/
And makes sense that someone tests it thoroughly first, which is probably outstanding, but I wanted to mention.

Comment (PromoFaux) [reply supportedos.txt:2]: mm, this is a good point.. the actual supported OS is controlled by a TXT record which we can change any time, so it doesn't rely on a release to the repo.

I have not fully tested 20.04 fully, but last time I did there was an issue with uploading debug logs

Comment (MichaIng) [reply supportedos.txt:2]: > the actual supported OS is controlled by a TXT record which we can change any time

Jep I recognised that, great Idea btw! This also makes it possible to halt updates in the very first place when a new release drops support for an old OS version that a user might run Pi-hole on currently.

**Pull request merged**